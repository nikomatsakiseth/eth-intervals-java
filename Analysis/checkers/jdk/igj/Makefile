
.SILENT:

# Variables
CHECKERS ?= ../..
CHECKERS_JAR ?= $(CHECKERS)/checkers.jar

LANGTOOLS ?= ../../../jsr308-langtools
JAVADOCLIB ?= $(LANGTOOLS)/build/classes
JAVAC ?= $(LANGTOOLS)/dist/bin/javac -Xlint
JAVAP ?= $(LANGTOOLS)/dist/bin/javap

RTJAR ?= $(JAVA_HOME)/jre/lib/rt.jar

clean:
	rm -rf build annotated jdk.jar

build:
	mkdir -p build


############## Validate annotated jdk against missing methods ##############
SOURCES = $(shell find src/java -name '*.java')
CLASSES = $(patsubst src/%.java,build/%.class, $(SOURCES))
DIFFS = $(patsubst src/%.java,build/%.diff, $(SOURCES))

.PHONY: classes
classes: $(SOURCES)
	echo "building the jdk..."
	mkdir -p build
	$(JAVAC) -cp $(CHECKERS_JAR):$(JAVADOCLIB) -d build $?

build/%.class: src/%.java
	$(JAVAC) -sourcepath src -cp $(CHECKERS_JAR):$(JAVADOCLIB) -d build $<

build/%.expected: build/%.class
	$(JAVAP) -cp $(JAVADOCLIB) $(subst /,.,$*)  | grep public | sed 's/ native / /g' | sed 's/? extends java\/lang\/Object/?/g' | sort > $@

build/%.actual: build/%.class
	$(JAVAP) -Xbootclasspath/p:build $(subst /,.,$*)  | grep public | sed 's/ native / /g' | sed 's/? extends java\/lang\/Object/?/g' | sort > $@

build/%.diff: build/%.expected build/%.actual
	echo "validating class:" $(subst /,.,$*)
	diff $? > $@

validate: classes $(DIFFS)

############## Annotate the annotated jdk file ##############
JAIFS = $(patsubst src/%.java,build/%.jaif, $(SOURCES))
ANNOTATED = $(patsubst src/%.java,annotated/%.class, $(SOURCES))

build/%.jaif: build/%.class
	extract-annotations $<

stubs: $(JAIFS)

annotated/%.class: build/%.jaif
	cd annotated; unzip $(RTJAR) $*.class > /dev/null
	insert-annotations $@ $<

annotated: $(ANNOTATED)

jdk.jar: annotated
	cd annotated; jar cf ../jdk.jar `find . -name '*.class'`

jdk: jdk.jar
