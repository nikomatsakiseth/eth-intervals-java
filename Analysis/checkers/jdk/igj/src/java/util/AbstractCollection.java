package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractCollection<E> implements java.util.Collection<E> {
  protected AbstractCollection() @ReadOnly {}
  public abstract @I java.util.Iterator<E> iterator() @ReadOnly;
  public abstract int size() @ReadOnly;
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean contains(java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public java.lang.Object[] toArray() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public <T> T @Mutable [] toArray(T @Mutable [] a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean remove(java.lang.Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean containsAll(@ReadOnly java.util.Collection<?> a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean addAll(@ReadOnly java.util.Collection<? extends E> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean removeAll(@ReadOnly java.util.Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean retainAll(@ReadOnly java.util.Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public java.lang.String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
