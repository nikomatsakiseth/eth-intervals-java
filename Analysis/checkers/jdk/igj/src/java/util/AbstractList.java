package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractList<E> extends @I java.util.AbstractCollection<E> implements @I java.util.List<E> {
  protected AbstractList() @ReadOnly {}
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public abstract E get(int a1) @ReadOnly;
  public E set(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public void add(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove(int a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public int indexOf(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int lastIndexOf(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(int a1, @ReadOnly java.util.Collection<? extends E> a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public @I java.util.Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.ListIterator<E> listIterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.ListIterator<E> listIterator(int a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.List<E> subList(int a1, int a2) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean equals(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
