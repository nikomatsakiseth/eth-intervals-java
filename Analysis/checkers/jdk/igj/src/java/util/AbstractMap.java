package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractMap<K, V> implements @I java.util.Map<K, V> {
  @I
  public static class SimpleEntry<K, V> implements @I java.util.Map.Entry<K, V>, @I java.io.Serializable {
      private static final long serialVersionUID = 0L;
    public SimpleEntry(K a1, V a2) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public SimpleEntry(@I java.util.Map.Entry<? extends K, ? extends V> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public K getKey() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V getValue() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V setValue(V a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public boolean equals(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
    public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public java.lang.String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
  }

  @Immutable
  public static class SimpleImmutableEntry<K, V> implements @Immutable java.util.Map.Entry<K, V>, @Immutable java.io.Serializable {
      private static final long serialVersionUID = 0L;
    public SimpleImmutableEntry(K a1, V a2) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public SimpleImmutableEntry(java.util.Map.Entry<? extends K, ? extends V> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public K getKey() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V getValue() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V setValue(V a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public boolean equals(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
    public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public java.lang.String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
  }

  protected AbstractMap() @ReadOnly {}
  public int size() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean containsValue(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean containsKey(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public V get(java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public V put(K a1, V a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public V remove(java.lang.Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void putAll(@ReadOnly java.util.Map<? extends K, ? extends V> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public @I java.util.Set<K> keySet() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.Collection<V> values() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public abstract @I java.util.Set<@I java.util.Map.Entry<K, V>> entrySet() @ReadOnly;
  public boolean equals(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public java.lang.String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
