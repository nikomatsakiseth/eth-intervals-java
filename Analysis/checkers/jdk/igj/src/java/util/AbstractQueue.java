package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractQueue<E> extends @I java.util.AbstractCollection<E> implements @I java.util.Queue<E> {
  protected AbstractQueue() @ReadOnly {}
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove() @Mutable { throw new RuntimeException("skeleton method"); }
  public E element() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(@ReadOnly java.util.Collection<? extends E> a1) @Mutable { throw new RuntimeException("skeleton method"); }
}
