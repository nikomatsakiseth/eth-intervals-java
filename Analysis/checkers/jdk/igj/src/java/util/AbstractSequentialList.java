package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractSequentialList<E> extends @I java.util.AbstractList<E> {
  protected AbstractSequentialList() @ReadOnly {}
  public E get(int a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E set(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public void add(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove(int a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(int a1, @ReadOnly java.util.Collection<? extends E> a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public @I java.util.Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public abstract @I java.util.ListIterator<E> listIterator(int a1) @ReadOnly;
}
