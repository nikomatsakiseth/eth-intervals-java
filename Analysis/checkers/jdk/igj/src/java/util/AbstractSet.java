package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractSet<E> extends @I java.util.AbstractCollection<E> implements @I java.util.Set<E> {
  protected AbstractSet() @ReadOnly {}
  public boolean equals(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean removeAll(@ReadOnly java.util.Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
}
