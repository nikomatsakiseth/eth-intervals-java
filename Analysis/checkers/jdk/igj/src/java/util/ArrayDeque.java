package java.util;
import checkers.igj.quals.*;

@I
public class ArrayDeque<E> extends @I java.util.AbstractCollection<E> implements @I java.util.Deque<E>, @I java.lang.Cloneable, @I java.io.Serializable {
    private static final long serialVersionUID = 0L;
  public ArrayDeque() @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayDeque(int a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayDeque(@ReadOnly java.util.Collection<? extends E> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public void addFirst(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void addLast(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offerFirst(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offerLast(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E removeFirst() @Mutable { throw new RuntimeException("skeleton method"); }
  public E removeLast() @Mutable { throw new RuntimeException("skeleton method"); }
  public E pollFirst() @Mutable { throw new RuntimeException("skeleton method"); }
  public E pollLast() @Mutable { throw new RuntimeException("skeleton method"); }
  public E getFirst() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E getLast() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peekFirst() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peekLast() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean removeFirstOccurrence(@ReadOnly java.lang.Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean removeLastOccurrence(@ReadOnly java.lang.Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offer(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove() @Mutable { throw new RuntimeException("skeleton method"); }
  public E poll() @Mutable { throw new RuntimeException("skeleton method"); }
  public E element() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peek() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void push(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E pop() @Mutable { throw new RuntimeException("skeleton method"); }
  public int size() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I java.util.Iterator<E> descendingIterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean contains(@ReadOnly java.lang.Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean remove(@ReadOnly java.lang.Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public java.lang.Object[] toArray() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public <T> T[] toArray(T[] a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I("N") ArrayDeque<E> clone() { throw new RuntimeException("skeleton method"); }
}
