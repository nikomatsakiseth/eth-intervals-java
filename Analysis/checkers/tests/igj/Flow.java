import checkers.igj.quals.*;
import java.util.*;

public class Flow {
    <T> @Immutable List<T> emptyList() { return null; }

    public void testFlow() {

        @Mutable List<String> m = emptyList();   // error

        List<String> im = emptyList();
        im.add("m"); // error
    }
}
