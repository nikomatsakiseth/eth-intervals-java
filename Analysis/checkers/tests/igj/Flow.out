:9: (type.incompatible)
:12: (method.invocation.invalid)
